File: BST Maximum Difference.cpp
class Solution{
public:
    Node *search(Node *root,int target,int &ans){
        if(root==NULL) return NULL;
        if(root->data==target){
            return root;
        }
        ans+=root->data;
        if(target>root->data){
            return search(root->right,target,ans);
        }
        return search(root->left,target,ans);
    }
    int dfs(Node *node){
        int res=node->data;
        int mn=1e9;
        if(node->left) mn=min(mn,dfs(node->left));
        if(node->right) mn=min(mn,dfs(node->right));
        if(mn==1e9) return res;
        return res+mn;
    }
    int maxDifferenceBST(Node *root,int target){
        int ans=0;
        Node *node=search(root,target,ans);
        if(node==NULL) return -1;
        ans-=dfs(node)-node->data;
        return ans;
    }
};

File: BST Maximum Difference.py
class Solution:
    def search(self,root,target,ans):
        if not root:
            return None
        if root.data==target:
            return root
        self.ans+=root.data
        if target>root.data:
            return self.search(root.right,target,ans)
        else:
            return self.search(root.left,target,ans)
    def dfs(self,node):
        res=node.data
        mn=10**9
        if node.left:
            mn=min(mn,self.dfs(node.left))
        if node.right:
            mn=min(mn,self.dfs(node.right))
        if mn==10**9:
            return res
        else:
            return res+mn
            
    def maxDifferenceBST(self,root, target):
        self.ans=0
        node=self.search(root,target,self.ans)
        if node==None:
            return -1
        self.ans-=self.dfs(node)-node.data
        return self.ans
